### PR TITLE
[IntroCSPy-week2] - Completed Final Exercises in Week2

### PR DESCRIPTION
**Pull request by kayleabox**
This PR has the final exercises for week2 of the class.

To run files in solutions directory: 
```
for f in solutions/*.py; do python3 "$f"; done
```

To run ccbill_project:
```
for f in solutions/ccbill_project/*.py; do python3 "$f"; done
```

To run guess_number: 
```
python3 solutions/guess_number/guess_number.py
```

### CODE REVIEW

File: week2/solutions/ccbill_project/cal_min_pay.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+balance at the end of the year in the format:
+
+Monthly interest rate= (Annual interest rate) / 12.0
+Minimum monthly payment = (Minimum monthly payment rate) x (Previous balance)
+Monthly unpaid balance = (Previous balance) - (Minimum monthly payment)
+Updated balance each month = (Monthly unpaid balance) + (Monthly interest
+rate x Monthly unpaid balance)
+'''
+def cal_min_pay(balance, annual_interest_rate, monthly_payment_rate):
+  monthly_interest = annual_interest_rate/12
```
Comment (inchoate) [inline week2/solutions/ccbill_project/cal_min_pay.py:18]: Space around the `/`.

File: week2/solutions/ccbill_project/cal_min_pay.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  remain_balance =  balance
```
Comment (inchoate) [inline week2/solutions/ccbill_project/cal_min_pay.py:19]: Just call it 'remaining_balance`.  :-)

File: week2/solutions/ccbill_project/cal_min_pay.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+So your program only prints out one thing: the remaining 
+balance at the end of the year in the format:
+
+Monthly interest rate= (Annual interest rate) / 12.0
+Minimum monthly payment = (Minimum monthly payment rate) x (Previous balance)
+Monthly unpaid balance = (Previous balance) - (Minimum monthly payment)
+Updated balance each month = (Monthly unpaid balance) + (Monthly interest
+rate x Monthly unpaid balance)
+'''
+def cal_min_pay(balance, annual_interest_rate, monthly_payment_rate):
```
Comment (inchoate) [inline week2/solutions/ccbill_project/cal_min_pay.py:17]: This isn't a useful function name. What is `cal`? What is `pay`?

Can you find a better name for this function that better describes what it actually does?

File: week2/solutions/ccbill_project/pay_one_year.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  print('Lowest Payment: ' + str(round(lowest_pay, 2)))
+
+def get_fixed_rate(monthly_payment, balance, monthly_interest):
+  remain_balance = balance
+
+  for month in range(12):
+    unpaid = remain_balance - monthly_payment 
+    remain_balance = unpaid + (monthly_interest * unpaid)
+    
+  if(remain_balance <= 0):
```
Comment (inchoate) [inline week2/solutions/ccbill_project/pay_one_year.py:25]: In Python we usually don't use `()` around conditionals. The Pythonic way would be:
```
if remain_balance <= 0:
```

File: week2/solutions/ccbill_project/pay_one_year_bi.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+Monthly payment upper bound = (Balance x (1 + Monthly interest rate)12) / 12.0
+
+Write a program that uses these bounds and bisection search to find the smallest monthly 
+payment to the cent such that we can pay off the debt within a year. Try it out with 
+large inputs, and notice how fast it is. Produce the same return value as you did in Problem 2.
+
+Note that if you do not use bisection search, your code will not run - your code only has 30 seconds to run on our servers.
+'''
+
+def pay_one_year_bi(balance, annual_interest_rate):
```
Comment (inchoate) [inline week2/solutions/ccbill_project/pay_one_year_bi.py:48]: `bi`?

File: week2/solutions/ccbill_project/pay_one_year_bi.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  upper_bound = (balance * (1 + monthly_interest)**12) /12.0
+  lower_bound = balance / 12
+  lowest_pay = get_fixed_rate(balance, monthly_interest, upper_bound, lower_bound)
+  print('Lowest Payment: ' + str(round(lowest_pay, 2)))
+
+def get_fixed_rate(balance, monthly_interest, upper_bound, lower_bound):
+  monthly_payment = round(lower_bound + (upper_bound - lower_bound)/2, 4)
+  remain_balance = check_monthly_payment(balance, monthly_interest, monthly_payment)
+
+  if(remain_balance <= 0 and remain_balance >= -0.01):
```
Comment (inchoate) [inline week2/solutions/ccbill_project/pay_one_year_bi.py:59]: Same:
```
if remain_balance <= 0 and remain_balance >= -0.01:
```

Comment (inchoate) [reply week2/solutions/ccbill_project/pay_one_year_bi.py:59]: Also, when I'm trying to check whether a number is in a range like you're doing here, I order the params like this to actually put the number between the extents of the range:
```
if -0.01 <= remain_balance and remain_balance <= 0:
```

This way it's clearer to me what I"m attempting to do.

File: week2/solutions/polysum.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+The perimeter of a polygon is: length of the boundary of the polygon
+Write a function called polysum that takes 2 arguments, n and s.
+This function should sum the area and square of the perimeter of the regular polygon.
+The function returns the sum, rounded to 4 decimal places.
+"""
+
+import math
+
+def polysum(n, s):
+  area = (0.25*n*s**2)/(math.tan(math.pi/n))
```
Comment (inchoate) [inline week2/solutions/polysum.py:14]: Math code is nearly always ugly code. :-/

**Review by inchoate — COMMENTED** [review]
A few comments.

File: week2/solutions/polysum.py (hunk 7 above)
Comment (kayleabox) [reply week2/solutions/polysum.py:14]: I can add more descriptive terms for the parameters. Change n to number_sides and s to side_length possibly.

**Pull request merged**